cpu/qn908x/periph/flexcomm.c

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (C) 2020 iosabi
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup     cpu_qn908x
+ *
+ * @{
+ *
+ * @file
+ * @brief       Flexcomm interrupt dispatch.
+ *
+ * @author      iosabi <iosabi@protonmail.com>
+ *
+ * @}
+ */
+
+#include <stdint.h>
+#include "cpu.h"
+#include "periph_conf.h"
+#include "vectors_qn908x.h"
+#include "flexcomm.h"
+
+#include "vendor/drivers/fsl_clock.h"
+
+#include "debug.h"
+
+int flexcomm_init(FLEXCOMM_Type *dev, flexcom_pselid_t mode)
+{
+    static const clock_ip_name_t flexcomm_clocks[] = FLEXCOMM_CLOCKS;
+    int flexcomm_num = flexcomm_instance_from_addr(dev);
+
+    if (flexcomm_num < 0 || flexcomm_num >= (int)ARRAY_SIZE(flexcomm_clocks)) {
+        DEBUG("Invalid flexcomm_num: %d\n", flexcomm_num);
+        return -1;
+    }
+    CLOCK_EnableClock(flexcomm_clocks[flexcomm_num]);
+    /* Reset the flexcomm. */
+    SYSCON->RST_SW_SET = 1u << flexcomm_num;
+    SYSCON->RST_SW_CLR = 1u << flexcomm_num;
+
+    /* Check that the mode is present in the FLEXCOMM.
+     * Bits 4, 5 and 6 tell whether the UART, SPI and I2C respectively are
+     * present. */
+    if ((dev->PSELID & (1u << (mode + 3))) == 0) {
+        DEBUG("Mode %d not present in FLEXCOMM%d\n", (int)mode, flexcomm_num);
+        return -1;
+    }
+
+    /* This also locks the peripheral so it can't be changed until device or
+     * peripheral is reset. */
+    dev->PSELID = (dev->PSELID & ~FLEXCOMM_PSELID_PERSEL_MASK) |
+                  FLEXCOMM_PSELID_LOCK_MASK |
+                  FLEXCOMM_PSELID_PERSEL(mode);
+    return flexcomm_num;
+}
+
+int flexcomm_instance_from_addr(FLEXCOMM_Type *dev)
+{
+    static const FLEXCOMM_Type *flexcomm_addrs[] = FLEXCOMM_BASE_PTRS;
+
+    for (uint8_t i = 0; i < ARRAY_SIZE(flexcomm_addrs); i++) {
+        if (flexcomm_addrs[i] == dev) {
+            return i;
+        }
+    }
```

Comment:
if they are evenly spaced, you can also just do `return dev - FLEXCOMM_BASE_PTRS` - could then even be a `static inline flexcomm_instance_from_addr()`

Reply:
```
#define FLEXCOMM0_BASE                           (0x40083000u)
#define FLEXCOMM1_BASE                           (0x40086000u)
#define FLEXCOMM2_BASE                           (0x40087000u)
#define FLEXCOMM3_BASE                           (0x4008F000u)
```

they are unfortunately not. I could play some tricks with address and LUT instead of branches like here (the loop is unrolled anyway). It is not that time critical since on the driver is only called on `*_init()` (whenever I need the instance it will be already available without lookup like in the isr_flexcomm call)